frameworks/keyed/solid-signals/src/main.jsx

```diff
@@ -1,109 +1,113 @@
+  colours = ["red", "yellow", "blue", "green", "pink", "brown", "purple", "brown", "white", "black", "orange"],
+  nouns = ["table", "chair", "house", "bbq", "desk", "car", "pony", "cookie", "sandwich", "burger", "pizza", "mouse", "keyboard"];
+
+function _random (max) { return Math.round(Math.random() * 1000) % max; };
+
+function buildData(count) {
+  let data = new Array(count);
+  for (let i = 0; i < count; i++) {
+    const [label, setLabel] = createSignal(`${adjectives[_random(adjectives.length)]} ${colours[_random(colours.length)]} ${nouns[_random(nouns.length)]}`);
+    data[i] = {
+      id: idCounter++,
+      label, setLabel
+    }
+  }
+  return data;
+}
+
+const Button = ({ id, text, fn }) =>
+  <div class='col-sm-6 smallpad'>
+    <button id={ id } class='btn btn-primary btn-block' type='button' onClick={ fn }>{ text }</button>
+  </div>
+
+const App = () => {
+  let rowId;
+  const [data, setData] = createSignal([]),
+    [selected, setSelected] = createSignal(null, (a, b) => a === b),
+    applySelection = selectWhen(selected, 'danger');
+
+  return <div class='container'>
+    <div class='jumbotron'><div class='row'>
+      <div class='col-md-6'><h1>SolidJS Keyed</h1></div>
+      <div class='col-md-6'><div class='row'>
+        <Button id='run' text='Create 1,000 rows' fn={ run } />
+        <Button id='runlots' text='Create 10,000 rows' fn={ runLots } />
+        <Button id='add' text='Append 1,000 rows' fn={ add } />
+        <Button id='update' text='Update every 10th row' fn={ update } />
+        <Button id='clear' text='Clear' fn={ clear } />
+        <Button id='swaprows' text='Swap Rows' fn={ swapRows } />
+      </div></div>
+    </div></div>
+    <table class='table table-hover table-striped test-data'><tbody>
+      <For each={ data() } transform={ applySelection }>{ row => (
+        rowId = row.id,
+        <tr model={ rowId }>
+          <td class='col-md-1' textContent={ rowId } />
+          <td class='col-md-4'><a onClick={ select } textContent={ row.label() } /></td>
+          <td class='col-md-1'><a onClick={ remove }><span class='glyphicon glyphicon-remove' aria-hidden="true" /></a></td>
+          <td class='col-md-6'/>
+        </tr>
+      )}</For>
+    </tbody></table>
+    <span class='preloadicon glyphicon glyphicon-remove' aria-hidden="true" />
+  </div>
+
+  function select(e, id) { setSelected(id); }
+
+  function remove(e, id) {
+    const d = data();
+    d.splice(d.findIndex(d => d.id === id), 1);
+    setData(d);
+  }
+
+  function run() {
+    freeze(() => {
+      setData(buildData(1000));
+      setSelected(null);
+    });
+  }
+
+  function runLots() {
+    freeze(() => {
+      setData(buildData(10000));
+      setSelected(null);
+    });
+  }
+
+  function add() { setData(data().concat(buildData(1000))); }
+
+  function update() {
+    freeze(() => {
+      const d = data();
+      let index = 0;
+      while (index < d.length) {
+        d[index].setLabel(d[index].label() + ' !!!');

```

Comment:
Is that the fine-grained strategy to you always talking about?^^

Reply:
lol

Reply:
Part of it. Fine grained reactive libraries are based on wrapping data access with getters and setters (either functions, object getter/setters, or proxies). So data is trees are broken into reactive atoms. Functions are the classic approach used found in libraries like Knockout, Ember etc and are still the most performant approach. i've written an introduction to the topic: [Finding Fine Grained Reactive Programming](https://levelup.gitconnected.com/finding-fine-grained-reactive-programming-89741994ddee?source=friends_link&sk=31c66a70c1dce7dd5f3f4229423ad127).

Proxies are more common these days being used by MobX and Vue 3 (keep in mind these latter 2 feed into VDOM so while change detection is fine-grained rendering is not). Proxies have taken a while to get browser support so I suspect we will see more in the future. But they definitely come with tradeoffs. Most in lack of explicitness. They hide the fact especially on update that change is happening, that you are dealing with anything other than a POJO. Personally it is why Solid has taken a very different approach to proxies than every other library by choosing not to use simple assignment on purpose. To understand that better I've written: [What Every JavaScript Framework Could Learn from React](https://medium.com/@ryansolid/what-every-javascript-framework-could-learn-from-react-1e2bbd9feb09?source=friends_link&sk=75b3f6f90eecc7d210814baa2d5ab52c)

Reply:
That is an explicit and absolutely non-generic call ... and please stop posting your articles, that will doesn't change that fact.

Reply:
No you are incorrect. In a sentence: you are pointing at a syntax difference that doesn't dictate either way whether the approach is data driven or not.

But I suppose it's a cop-out a bit to just point you to articles. It's a decent chunk of reading, and I was doing it just to hopefully save myself from explaining concepts again. But I will attempt again, under the assumption you are interested in understanding. I hope by the end of this you will understand the subtle differences which is why your posts gained some heat.

Ok first consider the differences between these:
```js
data = [/*... some data */]

// 1
for(let i = 0; i < data.length; i++) data[i] += 10;

//2
data.forEach((_, i) => data[i] += 10);

//3
function add10(list) { return data.forEach((_, i) => data[i] += 10));
add10(data);
```
The 1st is our typical for loop. The 2nd method is mechanically similar but now it is being called in a function form. The 3rd is slightly different in that the predicate is no longer in the callers control. `add10` might be incredibly useful in your domain and used everywhere but it is less general. Pretty straight forward. Now Lets consider the following:

```js
// 1
function action1() { doUpdate(); }

//2
function action2() { emitter.trigger("update"); }
// somewhere else
emitter.addListener("update", doUpdate);
```
Again these will mechanically do very similar things once wired up. But the subtle difference is the inversion of control. The coupling is looser and a different party could be responsible for wiring it up. A similar conversation comes up when comparing 1 vs 2 way binding. In React you might write(although you would probably use defaultValue):
```jsx
<input type="text" value={state.value} onInput={e => setState({ value: e.target.value })} />
```
This is still a full cycle equivalent to 2 way binding with 2 differences: It is much more verbose to wire up, and the flow is much more explicit. You see(and control) the direction data is travelling. This is known as uni-directional flow.

Finally, consider the difference between Proxies and function calls:
```js
function getValue(key) { return data[key]; }
function setValue(key, value) { data[key] = value; }

const pData = new Proxy(data, {
  get(target, key) { return target[key] },
  set(target, key, value) {
    target[key] = value;
  }
});

setValue("name", "John");
const name1 = get("name");

pData.name = "John;
const name2 = pData.name;
```
As you can see again mechanically accomplishing the same thing. There however is one key difference. The functions have the ability to be managed separately. I can pass around the getter or setter independently. And yes I can do that with the proxy by wrapping it in 2 functions. But my point is like the example above about 2 way binding this is just a syntax choice that has data flow repercussions although mechanically they can work the same.

---------------
Ok tying this all together. What is the difference here between the highlighted code snippet and Mikado's swap? Consider the first 2 examples where we have the different versions of the loop, and where we do direct updates versus registering an event emitter. Using the same language here is how Solid and Mikado differ on say Swap:

```js
// mikado
swapDOM(index1, index2) { 
  // ... do DOM work
}
swapDOM(1, 998);

// Solid
updateIndex(index1, value) {
  emitter[index1].trigger("update", value)
}

const data1 = data[1];
const data2 = data[998];
updateIndex(1, data1);
updateIndex(998, data2);

// somewhere else
data.forEach((_, i) => emitter[i].register("update", () => /* do DOM work */));
```
Even if on a different granularity than say a Virtual DOM library Solid's approach is detached from the updates. The data is independently managed than any DOM updates. You can use the data without ever touching the DOM, or you bind the same data to 10 views. Once wired yes updates can happen rather directly but the approach opens the door for the data to drive any update you want. Where as Mikado's helper method is just a glorified wrapper of direct DOM updates. The end user does not set the predicate so to speak (from our first example). It isn't just about reconciling or diffing approach, but that you control the application through the flow of data.

Now not all libraries need to be Data Driven. But this is what I mean by data driven and it checks with definition of the Virtual DOM libraries. They don't use a number of event emitters but they still update the data with no mindful consequence of what the DOM updates may be. The data update function doesn't know, the consuming template does.

Finally, since Proxies are just a syntax wrapper over functions. I could choose to use a Proxy to wrap either approach. I can have my set handler directly update the DOM or have it trigger registered updates. There is nothing inherent about it being a proxy that makes things different. Nothing makes using Proxies more pure or not. What makes something Data Driven is not its ability to tightly bind to DOM updates. No one would call jQuery Data Driven. Rather that conceptually updates are made to Data rather than the DOM. That the Data and DOM updates are not implicitly tightly coupled. That the data can update the DOM as desired without the person updating the data needing to be conscious of the DOM updates.

Reply:
Thanks for explanation, but I know these things already. By the way IoC is something I use very often, when it fits to the given task. Your example how mikado does apply swap is wrong for the proxy version. The truth is it looks very similiar to the approach you posted below. There is just one detail which makes my proxy implementation special, maybe unique, it completely derives from data to provide this: "That the data can update the DOM as desired without the person updating the data needing to be conscious of the DOM updates". If you find something similar, please show me, I'm interested in.

Reply:
BTW, bi-directional binding is a complete different topic ...

Reply:
> BTW, bi-directional binding is a complete different topic ...

Definitely. It is just a good example that we would all be familiar with where something effectively ends up being the same thing but the concept/spirit of the thing is very different. With direct updates or triggered, the very same code could more or less be run the same but there are fundamental differences on how you approach it.

And I mean I haven't checked but I assumed your Proxy implementation was different than the original Mikado implementation in that way. I just wanted to spell out how Solid or the new Proxy implementation is different from that.

> There is just one detail which makes my proxy implementation special, maybe unique, it completely derives from data to provide this: "That the data can update the DOM as desired without the person updating the data needing to be conscious of the DOM updates".

This is what I don't understand. I believe most implementations do this for the most part. Maybe comment on that in the other thread.